### ISSUE TITLE
[runtimes] Incorrect applied rotation

### ISSUE THREAD

**Issue by jabuwu** [EsotericSoftware/spine-runtimes#2158]
Hey there, I'm having a problem where Spineboy's feet become distorted when applying an alpha value between 0 and 1. The problem worsens over time as the alpha value changes.

<img width="432" alt="Screen Shot 2022-09-22 at 10 44 36 PM" src="https://user-images.githubusercontent.com/46233424/191899479-e1be1db5-98f9-4f2c-98e9-c2191efca102.png">

I have a fairly simple repro of this happening:
https://github.com/jabuwu/spine_rotation_bug/blob/main/src/main-simple.c
The `printf` at the end prints `-8806.835938` for the `arotation`.

You can see the bug visually by pulling that repo and running the build commands in the readme.

It started with commit a64b79b498a8d83a88e237c4ad83553aab4f3003 and I confirmed it's still happening in aeacf6e5028c25a2f34a9718f2d677cda6de6a70.

By the way, I have managed to get Spine working in Rust by [transpiling the C runtime to Rust](https://github.com/jabuwu/rusty_spine/blob/main/src/c/spine_c.rs). I hope this is okay. I made sure to include the Spine Runtimes License Agreement in my project. [The transpiled code has the added benefit of working in wasm using Bevy!](https://jabuwu.github.io/rusty_spine/)

**Comment by NathanSweet** [comment]
Repro using reference runtime spine-libgdx:
```
TrackEntry entry = state.setAnimation(0, "run", true);
for (int i = 0; i < 700; ++i) {
	entry.alpha = (i % 200 > 100) ? 0.5f : 1f;
	state.update(0.016f);
	state.apply(skeleton);
	skeleton.updateWorldTransform();
}
entry.alpha = 0.5f;
state.update(0.016f);
state.apply(skeleton);
skeleton.updateWorldTransform();
for (int i = 0; i < 2000; ++i) {
	state.update(0.016f);
	state.apply(skeleton);
	skeleton.updateWorldTransform();
}
Bone front_foot = skeleton.findBone("front-foot-tip");
System.out.println(front_foot.arotation); // -8806.837
```

Please see the TrackEntry `alpha` docs:
http://esotericsoftware.com/spine-api-reference#TrackEntry-alpha

> Values < 1 mix this animation with the skeleton's current pose (usually the pose resulting from lower tracks). Defaults to 1, which overwrites the skeleton's current pose with this animation.
>
> Typically track 0 is used to completely pose the skeleton, then alpha is used on higher tracks. It doesn't make sense to use alpha on track 0 if the skeleton pose is from the last frame render.

Since you are applying only one animation with `alpha` < 1, you are mixing between the current pose, which is from a previous frame of your animation, and your animation over and over. That isn't really a sensible thing to do. Generally you want to pose the skeleton to some known values (some other animation or the setup pose), then apply an animation using `alpha` < 1.

As for why the rotation values get wacky, when AnimationState applies a rotate timelines with `alpha` < 1, it has to interpolate between the current pose rotation and the animation pose's rotation. There are two ways around a circle, so AnimationState needs to choose clockwise or counterclockwise. The first time it makes this decision for a TrackEntry, it uses the shortest rotation. After that, it continues to use that direction, even if during the animations it becomes the longer direction. It remembers the direction so that the bones don't jump to the other side of the circle.

In your test case, after applying the animations long enough, I assume sometimes the interpolation between the rotation of the two poses becomes large when the initially chosen direction becomes the long way around. That makes for an occasional large change to the interpolated rotation. Doing it a lot eventually gives really odd values.

**Comment by NathanSweet** [comment]
Your Rust runtime is really neat, thanks for sharing! Yes, of course it is totally fine to do as long as you include the Spine Runtimes License Agreement, as you have. Cheers!

**Comment by jabuwu** [comment]
@NathanSweet Thanks for the info! However, I believe there is still a problem. It's my fault for oversimplifying the "repro" case to such an extent but when I originally noticed the bug, I was doing exactly that. I had a track on 0 (spineboy's "idle") and alpha'd into the run animation. That's when I experienced this behavior.

I updated my repo to demonstrate that you get equally strange behavior when doing as you suggest. Unfortunately, the simple test case no longer demonstrates the problem, but see here:
https://github.com/jabuwu/spine_rotation_bug/blob/565a08ff65c4f1457e9d10d81ee02c9257436d3d/src/main.cpp#L134

The for loops are still necessary to "accelerate" the bug.

https://user-images.githubusercontent.com/46233424/192152733-a6f88e07-4e48-4320-8699-f49dbf5c0177.mov

This is the original behavior I was experiencing:

https://user-images.githubusercontent.com/46233424/192153004-971c147e-00f1-45d0-b12d-f7baa3e52ff7.mov

This is through transpiled code, so it's entirely possible the transpiler causes this, hence the C examples I've provided. The code for this is here:
https://github.com/jabuwu/bevy_spine/blob/main/examples/spineboy/player.rs

But, as I mentioned, the problem goes away in all of these cases if I go back to d9935741c2b84abea85e961489ed71b5b473aa64, one commit before a64b79b498a8d83a88e237c4ad83553aab4f3003

I might still be doing something wrong but I'm not sure what that is. I'm very sorry if that's the case. I should probably be using mix durations in animation state data for this transition instead, but still, I was surprised by this result.

Thanks for your time!

**Comment by NathanSweet** [comment]
Hmm. I think the bug has the same cause even with your second code listing: timelines are applied with TrackEntry `alpha` < `1.0` without something else setting the value, so the `alpha` interpolates between the last frame and the current frame.

When you apply the `idle` animation, it does not key `front-foot-tip`, so it does not set the rotation of that bone. Applying an animation only sets the properties that are keyed. Then when you apply the `run` animation you get the problem as described before.

One fix is to key all the same properties in `idle` that are keyed in any animations applied on top using `alpha`. `ctrl+shift+L` in the dopesheet for `Key Shown` can be useful: key shown in `run` after the last frame, cut those keys, paste into `run` after the last frame, go to frame 0 and `Key Shown`, then delete the keys pasted after the last frame.

Another fix is to reset everything back to the setup pose before applying the AnimationState by calling Skeleton `setToSetupPose` (or just `setBonesToSetupPose`). This ensures properties that are only keyed in `run` won't use values from the last frame. For an property like `front-foot-tip` bone rotation that is not keyed in `idle`, the result will be that `alpha` interpolates from the setup pose to the animation pose.

Sorry you ran into trouble with this, it is a bit tricky!

**Comment by jabuwu** [comment]
Unfortunately, even with a calls to `spSkeleton_setToSetupPose`, the problem persists the same as it did before. I've updated my repo once more: https://github.com/jabuwu/spine_rotation_bug/commit/fac28f0173bb2acaac6108a55c149ce197290b0f

**Comment by NathanSweet** [comment]
TL;DR: We fixed a nasty bug and it should be working correctly once the fix is pushed, even if results can be surprising at times.

After a lot of investigation: The problem occurs because of how AnimationState remembers the rotation direction that is initially chosen, as described above. Additionally, when a full rotation has been completed, AnimationState also needs to remember that, else the interpolation between rotations would jump around. Here's a [forum post](http://esotericsoftware.com/forum/Timeline-and-Animation-API-14504?p=63812#p63812) about it with some GIFs.

First we found an issue with how the full rotations were counted. Remember we're interpolating between rotations of two animations that very likely change over time. When the difference between the bones' rotations reverses directions and crosses, we need to decrement a full rotation. Previously we were incrementing a full rotation, causing the rotation values to skyrocket. Please enjoy this wonderful GIF (after our fix):

![](http://n4te.com/x/5515-loops.gif)

The blue and red lines represent bone rotations from the two animations. The green line is the interpolated rotation, 50% between the blue and red bones. It remains stable over time when the bones cross, reverse directions, and cross again. We'll commit the fix to the `4.2-beta` and link to this issue. It's too risky for `4.1`, which is intended to be stable, like all our non-beta releases.

With that fix out of the way, we still see strange behavior with your scenario, even though we no longer see astronomical rotation values. We made a simpler repro ([project](http://n4te.com/x/5518-exports.zip)). I believe it's the same issue but doing hundreds of `apply` calls was too hard to debug. First:

```
TrackEntry entry = state.setAnimation(0, "run", true);
entry.alpha = 0.5f;
state.update(0.015553951f);
state.apply(skeleton);
```
That gives the expected result when `alpha` is adjusted from 0 (setup pose) to 1 (animation pose):
![](http://n4te.com/x/5524-expected1.gif)
It looks a bit stupid but it's correct: the foot bends a fair amount clockwise. Next:
```
TrackEntry entry = state.setAnimation(0, "run", true);
entry.alpha = 0.5f;
state.update(0.015553951f);
state.update(0.15413499f); // <-- added this line
state.apply(skeleton);
```
![](http://n4te.com/x/5523-expected2.gif)
The foot moves counterclockwise slightly. Finally:
```
TrackEntry entry = state.setAnimation(0, "run", true);
entry.alpha = 0.5f;
state.update(0.015553951f);
state.apply(skeleton); // <-- added this line
state.update(0.15413499f);
state.apply(skeleton);
```
![](http://n4te.com/x/5525-surprise.gif)
This is actually correct! What happened? The first `apply` started moving the foot clockwise. The second `apply` is for a very different pose, but we have already established we're moving clockwise, so around we go the long way.

Note it's much easier to see what's happening when you adjust `alpha` from 0 to 1. If you just pick a pose in the middle, it can look wrong without being able to see why. Here is your reproduction case, helping to see why the feet are curled at 50% `alpha`:
![](http://n4te.com/x/5522-repro.gif)

Most mixing has a short duration and is done between relatively similar poses, so doesn't have this issue. If mixing for a long duration or otherwise any time you are mixing to a very different pose, you may want to reset the rotations that are stored for the TrackEntry using [resetRotationDirections](http://esotericsoftware.com/spine-api-reference#TrackEntry-resetRotationDirections).

Another option is to enable [shortestRotation](http://esotericsoftware.com/spine-api-reference#TrackEntry-shortestRotation). When rotation always goes the short way around, you can never have issues with the initially chosen rotation direction surprising you by rotating the long way around later. The downside is the shortest direction often changes mid-animation, causing the bones to flip to the other side.

**Comment by jabuwu** [comment]
Thanks for the explanation! I can appreciate how difficult it is to sort out these rotation issues. I can barely remember how to calculate the shortest angle half the time. I think my main concern was that the changes are already breaking compared to how 4.1 worked, but I also understand the original commit was fixing someone else's rotation issue.

In any case, I'm not shipping production code over here, I only want to get a simple Spineboy demo working in Rust. Setting shortest rotation does work for me! I tried to use resetRotationDirections before but I don't think it's available in the C runtime.

**Comment by badlogic** [comment]
Port https://github.com/EsotericSoftware/spine-runtimes/commit/3ea4f66642f9a82eece1f0e9c10b273373306ded

- [x] C#
- [x] C
- [x] C++
- [x] TS

**Comment by badlogic** [comment]
This is now implemented in the 4.2-beta branch. I've also added `spTrackEntry_resetRotationDirections()`. Thanks for reporting!
